cores/esp8266/core_esp8266_i2s.c

```diff
@@ -194,46 +219,94 @@ bool ICACHE_FLASH_ATTR i2s_write_lr(int16_t left, int16_t right){
   return i2s_write_sample(sample);
 }
 
+bool ICACHE_FLASH_ATTR i2s_write_lr_nb(int16_t left, int16_t right){
+  int sample = right & 0xFFFF;
+  sample = sample << 16;
+  sample |= left & 0xFFFF;
+  return i2s_write_sample_nb(sample);
+}
+
 //  END DMA
 // =========
 // START I2S
 
 
-static uint32_t _i2s_sample_rate;
+static uint32_t _i2s_sample_rate = 0;
 
 void ICACHE_FLASH_ATTR i2s_set_rate(uint32_t rate){ //Rate in HZ
   if(rate == _i2s_sample_rate) return;
   _i2s_sample_rate = rate;
-  uint32_t i2s_clock_div = (I2SBASEFREQ/(_i2s_sample_rate*32)) & I2SCDM;
-  uint8_t i2s_bck_div = (I2SBASEFREQ/(_i2s_sample_rate*i2s_clock_div*2)) & I2SBDM;
-  //os_printf("Rate %u Div %u Bck %u Frq %u\n", _i2s_sample_rate, i2s_clock_div, i2s_bck_div, I2SBASEFREQ/(i2s_clock_div*i2s_bck_div*2));
 
-  //!trans master, !bits mod, rece slave mod, rece msb shift, right first, msb right
-  I2SC &= ~(I2STSM | (I2SBMM << I2SBM) | (I2SBDM << I2SBD) | (I2SCDM << I2SCD));
-  I2SC |= I2SRF | I2SMR | I2SRSM | I2SRMS | ((i2s_bck_div-1) << I2SBD) | ((i2s_clock_div-1) << I2SCD);
+  // Find closest divider
+  uint32_t cpu_freq = ets_get_cpu_frequency() * 1000000L;
+  int bestfreq = 0;
+  uint32_t i2s_clkm_div, i2s_bck_div;
+
+  // CLK_I2S = CPU_FREQ / I2S_CLKM_DIV_NUM
+  // BCLK = CLK_I2S / I2S_BCK_DIV_NUM
+  // WS = BCLK/ 2 / (16 + I2S_BITS_MOD)
+  // Note that I2S_CLKM_DIV_NUM must be >5 for I2S data
+  // I2S_CLKM_DIV_NUM - 5 - 63
+  // I2S_BCK_DIV_NUM - 2 - 63
+  for (int bckdiv = 2; bckdiv < 64; bckdiv++) {
+    for (int clkmdiv = 5; clkmdiv < 64; clkmdiv++) {
+      uint32_t testfreq = cpu_freq / (bckdiv * clkmdiv * 32);
```

Comment:
maybe we just hard-code these with C++ template meta-programming, and only calculate those ones not hard-coded?

but anyway the loop can be finished within few ms, and it's not called really often.